pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go

```diff
@@ -0,0 +1,142 @@
+func NewCachingMiddleware(cachingService caching.CachingService) plugins.ClientMiddleware {
+	log := log.New("caching_middleware")
+	if err := prometheus.Register(QueryRequestHistogram); err != nil {
+		log.Error("error registering prometheus collector 'QueryRequestHistogram'", "error", err)
+	}
+	if err := prometheus.Register(ResourceRequestHistogram); err != nil {
+		log.Error("error registering prometheus collector 'ResourceRequestHistogram'", "error", err)
+	}
+	return plugins.ClientMiddlewareFunc(func(next plugins.Client) plugins.Client {
+		return &CachingMiddleware{
+			next:    next,
+			caching: cachingService,
+			log:     log,
+		}
+	})
+}
+
+type CachingMiddleware struct {
+	next    plugins.Client
+	caching caching.CachingService
+	log     log.Logger
+}
+
+// QueryData receives a data request and attempts to access results already stored in the cache for that request.
+// If data is found, it will return it immediately. Otherwise, it will perform the queries as usual, then write the response to the cache.
+// If the cache service is implemented, we capture the request duration as a metric. The service is expected to write any response headers.
+func (m *CachingMiddleware) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	if req == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	reqCtx := contexthandler.FromContext(ctx)
+	if reqCtx == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	// time how long this request takes
+	start := time.Now()
+
+	// First look in the query cache if enabled
+	hit, cr := m.caching.HandleQueryRequest(ctx, req)
+
+	defer func() {
+		// record request duration if caching was used
+		if ch := reqCtx.Resp.Header().Get(caching.XCacheHeader); ch != "" {
+			QueryRequestHistogram.With(prometheus.Labels{
+				"datasource_type": req.PluginContext.DataSourceInstanceSettings.Type,
+				"cache":           ch,
+				"query_type":      getQueryType(reqCtx),
+			}).Observe(time.Since(start).Seconds())
+		}
+	}()
+
+	// Cache hit; return the response
+	if hit {
+		return cr.Response, nil
+	}
+
+	// Cache miss; do the actual queries
+	resp, err := m.next.QueryData(ctx, req)
+
+	// Update the query cache with the result for this metrics request
+	if err == nil && cr.UpdateCacheFn != nil {
+		cr.UpdateCacheFn(ctx, resp)
+	}
+
+	return resp, err
+}
+
+// CallResource receives a resource request and attempts to access results already stored in the cache for that request.
+// If data is found, it will return it immediately. Otherwise, it will perform the request as usual. The caller of CallResource is expected to explicitly update the cache with any responses.
+// If the cache service is implemented, we capture the request duration as a metric. The service is expected to write any response headers.
+func (m *CachingMiddleware) CallResource(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
```

Comment:
This gets a bit weird/hard to follow now when (as I understand) the actual caching is made in api/plugin_resource.go, but the check to see if it's already cached is made here. That's not optimal.

Reply:
Agreed, but not sure how to do it better right now, given the sender reading is done upstream.

Would it make sense to leave a comment here pointing to the caching call in `plugin_resource.go`, and create a follow-on task as you suggested above?

Reply:
Yeah possibly

Reply:
I've left a comment [b7290fb](https://github.com/grafana/grafana/pull/65616/commits/b7290fbd6b7914bc2412b7f92465c5a28ba2220c)

And created a follow on issue https://github.com/grafana/grafana/issues/66297. Will prioritize this post-merge.